### PR TITLE
add load test for L4 LB nodesync

### PR DESCRIPTION
**Pull request by freehan**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

Add a generic load test for L4 LBs

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:
```docs

```

### CODE REVIEW

**Comment by mm4tt** [comment]
/assign @jkaniuk 

Jacek, could you please review this?

File: clusterloader2/pkg/measurement/common/loadbalancer_nodesync_latency.go
Comment (prameshj) [inline clusterloader2/pkg/measurement/common/loadbalancer_nodesync_latency.go:194]: Should we break here? 

Also, will events be in reverse chronological order? (so the latest lb update shows up first?). Otherwise we might catch the event where the LB was updated for the first time right?

File: clusterloader2/pkg/measurement/common/loadbalancer_nodesync_latency.go
Comment (prameshj) [inline clusterloader2/pkg/measurement/common/loadbalancer_nodesync_latency.go:239]: Maybe add a TODO to return an error here if the latency is higher than an upper bound (say 30 mins)?

File: clusterloader2/pkg/measurement/common/loadbalancer_nodesync_latency.go
Comment (prameshj) [inline clusterloader2/pkg/measurement/common/loadbalancer_nodesync_latency.go:113]: The measure step picks one candidate node, marks it to be excluded from LB and then measures the time taken for the update event to show up. 

This is a good test, but I think the most problematic scenarios are where all nodes are removed and replaced by a new set of nodes, causing discrepancy in the k8s and GCE layer until steady state. Can we trigger that scenario by doing kubectl delete of all nodes?

**Review by prameshj — COMMENTED** [review]
Logic looks good to me. Added some follow-up questions.

File: clusterloader2/testing/l4lb/service.yaml
Comment (jkaniuk) [reply clusterloader2/testing/l4lb/service.yaml:8]: We do not have to support 1.16-

File: clusterloader2/testing/l4lb/service.yaml
Comment (jkaniuk) [inline clusterloader2/testing/l4lb/service.yaml:15]: I think that 8080 port on a node will conflict with NodeLocal DNSCache

File: clusterloader2/testing/l4lb/service.yaml
Comment (jkaniuk) [inline clusterloader2/testing/l4lb/service.yaml:16]: Shouldn't it match `containerPort: 8080` ?

File: clusterloader2/testing/l4lb/config.yaml
Comment (jkaniuk) [inline clusterloader2/testing/l4lb/config.yaml:72]: nit: space missing

File: clusterloader2/testing/l4lb/config.yaml
Comment (jkaniuk) [inline clusterloader2/testing/l4lb/config.yaml:12]: The biggest clusters can handle 100 qps pod creation and 1/3 of that for deletion. Default controller-manager qps limit (for smaller clusters) is 20 (deletion also 1/3 of that)

**Review by jkaniuk — COMMENTED** [review]
Partially reviewed, mostly config.

@tosi3k - could you look into that?

File: clusterloader2/testing/l4lb/service.yaml
Comment (prameshj) [reply clusterloader2/testing/l4lb/service.yaml:15]: Good point, though it is only a problem if the pod is in "hostNetwork mode". This one is not.

Comment (jkaniuk) [reply clusterloader2/testing/l4lb/service.yaml:15]: Isn't port and targetPort swapped then? service should be on 80 (port), pod exposes 8080 (targetPort)

File: clusterloader2/pkg/measurement/common/loadbalancer_nodesync_latency.go
Comment (freehan) [reply clusterloader2/pkg/measurement/common/loadbalancer_nodesync_latency.go:194]: Adding a break. 

The measurement only triggers one nodesync event, so there should be only one nodeSyncEvent occur.

File: clusterloader2/pkg/measurement/common/loadbalancer_nodesync_latency.go
Comment (freehan) [reply clusterloader2/pkg/measurement/common/loadbalancer_nodesync_latency.go:113]: The problem with that is kubelet will try to add the node back immediately. The reason why I did not pick a rolling update or node pool operation is because it might cause multiple nodesyncs. Causing the unnecessary noise for measuring latency.

File: clusterloader2/pkg/measurement/common/loadbalancer_nodesync_latency.go
Comment (freehan) [reply clusterloader2/pkg/measurement/common/loadbalancer_nodesync_latency.go:239]: Ack

File: clusterloader2/testing/l4lb/config.yaml
Comment (freehan) [reply clusterloader2/testing/l4lb/config.yaml:12]: ack

File: clusterloader2/testing/l4lb/service.yaml
Comment (freehan) [reply clusterloader2/testing/l4lb/service.yaml:8]: done

File: clusterloader2/testing/l4lb/service.yaml
Comment (freehan) [reply clusterloader2/testing/l4lb/service.yaml:16]: good catch

File: clusterloader2/testing/l4lb/service.yaml
Comment (freehan) [reply clusterloader2/testing/l4lb/service.yaml]: yes. Corrected the yaml. 

Now service and container port are both 80

**Comment by freehan** [comment]
Ready for another round.

File: clusterloader2/pkg/measurement/common/loadbalancer_nodesync_latency.go
Comment (wojtek-t) [inline clusterloader2/pkg/measurement/common/loadbalancer_nodesync_latency.go:53]: Can we avoid this?

I see this is used only in one place for Search in events.
Can we instead create an appropriate field selector outselves?

It should actually be as simple as:
```
selector := client.CoreV1().Events(s.selector.Namespace).GetFieldSelector(svc.Name, svc.Namespace, nil, nil)
```

File: clusterloader2/pkg/measurement/common/loadbalancer_nodesync_latency.go
Comment (wojtek-t) [inline clusterloader2/pkg/measurement/common/loadbalancer_nodesync_latency.go:164]: break ?

File: clusterloader2/pkg/measurement/common/loadbalancer_nodesync_latency.go
Comment (wojtek-t) [inline clusterloader2/pkg/measurement/common/loadbalancer_nodesync_latency.go:185]: I'm not a fan of depending on events in a programatic way - events aren't guaranteed.

What if instead we would make the measurement GCE-specific and try to get TargetPool contents from GCE and detect if the node was added/removed based on this?

The event-based test is really implementation-specific - we should target something more blackbox.

Comment (freehan) [reply clusterloader2/pkg/measurement/common/loadbalancer_nodesync_latency.go:185]: Making a measurement using GCE API calls has a few downsides in this tests. The test has to loop thru a potential high number of LBs and make GCE API calls repeatedly until the expected state has reached. This would eat up the read RPS quota in GCE the project and potentially cause controllers to hit GCE RateLimiting errors. Having a higher interval would make the measurement less accurate when high number of LBs are used in the load test.


Yes. I agree depending on events are less straightforward. We brainstormed a few ways to make this measurement. One better alternative is to read a metrics exposed by kube-controller-manager directly on master, but cluster loader currently cannot reach master. While the event creation timestamp does provide an accurate latency measurement and it is more a blackbox test than checking TargetPool.

File: clusterloader2/pkg/measurement/common/loadbalancer_nodesync_latency.go
Comment (freehan) [reply clusterloader2/pkg/measurement/common/loadbalancer_nodesync_latency.go]: Good call.

File: clusterloader2/pkg/measurement/common/loadbalancer_nodesync_latency.go
Comment (freehan) [reply clusterloader2/pkg/measurement/common/loadbalancer_nodesync_latency.go:153]: Fixed

**Review by prameshj — APPROVED** [review]

File: clusterloader2/pkg/measurement/common/loadbalancer_nodesync_latency.go
Comment (wojtek-t) [inline clusterloader2/pkg/measurement/common/loadbalancer_nodesync_latency.go:174]: Continuing the discussion about events.

> Consuming GCE API calls quota.

How many services you would eventually like to test? I would like to understand if simple bump of quota won't help here. 
Instead of getting target pools one by one - can we list (potentially in batches)?


The current event-based approach is also inefficient, because given that your `s.selector.Namespace` is effectively empty, we will be listing all events from the whole clusters for every service independently (i.e. the returned result will be small, but we will need to internally process and filter out all events for every call).

Comment (freehan) [reply clusterloader2/pkg/measurement/common/loadbalancer_nodesync_latency.go:174]: I want to eventually test up to 500 LBs. (probably manual one time) And setup CI for 100 LBs. I have seen customers have >100 LBs. 

Listing TargetPools could not tell what instances are included. It requires a Get call to investigate the content and compare the instances one by one. 

I think clusterloader automatically creates separate namespaces for the load test. So `s.selector.Namespace` is only pointing to the test namespace. As I am using the feildSelector, the event Get call should only return events associated with the service, no?

Moreover, the event-based approach does not need to efficient, the retry interval can be configured to very high. This is because the latency measurement is depending on the `creationTimestamp` instead of the time when the validation took place.

Comment (wojtek-t) [reply clusterloader2/pkg/measurement/common/loadbalancer_nodesync_latency.go:174]: > I want to eventually test up to 500 LBs. (probably manual one time) And setup CI for 100 LBs. I have seen customers have >100 LBs.

OK - that doesn't sound that huge. On what size of cluster would you like to to test it?


> Listing TargetPools could not tell what instances are included. It requires a Get call to investigate the content and compare the instances against nodes one by one.

Got it.
I'm still a bit reluctant to using events - this is too whitebox-y to me. It's not future-proof, the fact that particular event was generated doesn't meen that LB works - it depends on the current implementation how the event is generated...

How sure are we that GCE API quota will be a problem? Can we mitigate it by choosing N random services and checking it for them? What if we would bump quotas?
[How many API calls do we need for updating nodes in a TP?]


> I think clusterloader automatically creates separate namespaces for the load test. 

Yes - but the selector.namespace is set to constant when you're parsing the config:
https://github.com/kubernetes/perf-tests/pull/1785/files#diff-dd38242128260b4a9cad8419f28cfb7f518618cd2e49c809fa532cfb4df36f4bR96
And you don't specify namespaces there - so it will be empty.
I'm not saying it wont work - I'm saying that it will be underneath listing all events in the system and filtering appropriate one.

> Moreover, the event-based approach does not need to efficient, the retry interval can be configured to very high. This is because the latency measurement is depending on the creationTimestamp instead of the time when the validation took place.

If we really need events (which I really hope we don't), a single informer sounds like a better option anyway:
https://github.com/kubernetes/perf-tests/blob/master/clusterloader2/pkg/measurement/util/informer/informer.go

Comment (freehan) [reply clusterloader2/pkg/measurement/common/loadbalancer_nodesync_latency.go:174]: > I'm still a bit reluctant to using events - this is too whitebox-y to me. It's not future-proof, the fact that particular event was generated doesn't mean that LB works - it depends on the current implementation how the event is generated...

But checking TargetPool is also very whitebox-y. It only works for NetLB. For the classic L4 ILB, the event based approach also works. 

> How sure are we that GCE API quota will be a problem? 

The GCE API read quota is per project and across all APIs. By default, the quota is 20 QPS. I think it can be bumped but I only have impression with it bumped to 40 QPS at most. For 500 TPs, it took at least (assuming 20 QPS), 25s to traverse all of them. 

> Can we mitigate it by choosing N random services and checking it for them? What if we would bump quotas? [How many API calls do we need for updating nodes in a TP?]

This particular measurement is measuring how long would it take for service controller to update all LBs in the cluster when a nodesync got triggered. So it has to scan thru all LBs to be accurate. 

For each TP update, it actually consists of multiple GCE API calls. 1 List Instance , 1 TargetPool Add instances, 1 TargetPool delete instances, 2 Get TargetPool. So it is quite a few calls already. 


> If we really need events (which I really hope we don't), a single informer sounds like a better option anyway:
https://github.com/kubernetes/perf-tests/blob/master/clusterloader2/pkg/measurement/util/informer/informer.go

I can change it to use informer if efficiency is the concern here.

Comment (wojtek-t) [reply clusterloader2/pkg/measurement/common/loadbalancer_nodesync_latency.go:174]: > But checking TargetPool is also very whitebox-y. It only works for NetLB. For the classic L4 ILB, the event based approach also works.

Can we use something different for ILB?
Maybe pinging the service and checking when it starts to reach the new node (we would have to add node instead of deleting it instead)?
I would really like to avoid event dependency...

>  For 500 TPs, it took at least (assuming 20 QPS), 25s to traverse all of them.

FYI - We're bumping it to 200qps in our largest projects.

Comment (prameshj) [reply clusterloader2/pkg/measurement/common/loadbalancer_nodesync_latency.go:174]: For ILB(Subsetting), I am planning to add a new measurement similar to [exec](https://github.com/kubernetes/perf-tests/blob/05b63b33bd15916ca92f9c724cc8ed83eca6a056/clusterloader2/pkg/measurement/common/exec.go#L47) so we can list the NEGs and verify the node names to be what we expect.

The presence of event can be used to trigger the NEG measurement, so we do not check the contents until necessary. We could omit it too. 

For IG-based ILB, we would need to list contents of the single instance group, since it is shared.
Perhaps we need very similar measurements for each case - NEG, IG, TargetPool. For IG and TargetPool, the steps are - 1) List all cluster nodes 2) list all instances in GCE resource. Ensure they have the same entries.

NEG is trickier since it uses a subset.

Comment (freehan) [reply clusterloader2/pkg/measurement/common/loadbalancer_nodesync_latency.go:174]: Let me summarize and compare the 2 approaches:

1. Event based approach (this PR)
   - Simple to implement
   - Latency measurement is accurate (from Nodesync starts and service controller finishes update LB hosts and emits event)
   - Can be used for both classic ILB/NetLB. 
   - Implementation specific (depending on the OSS service controller emitting the event)

2. TargetPool based approach
  - Polling based
  - Measurement is dependent on the frequency of the observation. (needs high RPS to )
  - Implementation specific (Only fits classic NetLB with Target Pool)


There are essentially 2 ways to implement the TargetPool based measurement:

A. Use Exec measurement as building block to run gcloud commands, the rough steps will be: 
  - Use the NetLB naming scheme and generate TargetPool name
  - Find the region
  - Run `gcloud to run gcloud compute target-pool describe ${targetpool} --region ${region}` (There is no list-istance gcloud API on targetPool)
  - Then do string parsing on the return output to match up with the expected nodes.

B. Introduce GCE API go clients into clusterloader2. 
- Construct the parameters (Targetpool Name and region) and then execute 
- use GCE API go client to poll target pool instances and parse it via golang. 

A few more factors to consider:
1. The immediate goal of this load test is to measure the scalability improvements caused by the parallelized backend management. 
2. RegionalBackendService migration is going to kick start soon. New NetLBs will be based on RBS instead of target pools

Overall, I would suggest the event based approach is a good compromise before RBS is in place. And we can improve this measurement incrementally. WDYT?

**Comment by Liu-Aohan** [comment]
Hi,

Your Github profile and the contributions are really impressive and the work in Kubernetes is exactly what we are looking for!

As Shopee is expanding our Engineering teams here in Singapore and looking for talents like you. Here's our current openings https://careers.shopee.sg/job-detail/2628/ and I would like to check would that be something interesting to you?

Let me know if you are open for a quick discussion and I'm more than happy to assist you. If you're not keen right now, I'd still love to connect and stay in touch for other exciting opportunities in future.

Here are some key stats of Shopee for your quick review:

- 6th year upon launching
- Global presence in 11 countries, 13 offices
- 1 billion total # plays for in-app games
- 429.8 million gross orders just for Q1 2020
- US$6.2 billion gross merchant value just for Q1 2020  

Please feel free to add my WeChat: 13733199078 and ask any questions. Hope to hear from you soon!

Best regards

File: clusterloader2/pkg/measurement/common/loadbalancer_nodesync_latency.go
Comment (wojtek-t) [reply clusterloader2/pkg/measurement/common/loadbalancer_nodesync_latency.go:174]: I'm definitely not  fan of relying on events, but I don't want to block this infinitely.

So for the sake of making progress, let's proceed with the event-based approach for now (to unblock your tests).
I would like to get rid of this eventually, but we may discuss that separately.

So let's just fix it to use informer instead of periodically listing and proceed with this.

**Comment by freehan** [comment]
Changed to use Event informer for processing. Ready for another round

File: clusterloader2/pkg/measurement/common/loadbalancer_nodesync_latency.go
Comment (freehan) [reply clusterloader2/pkg/measurement/common/loadbalancer_nodesync_latency.go:233]: I added a waitTimeout in the measurement. The test would fail if it took more than the timeout to measure nodesync.

**Comment by wojtek-t** [comment]
@jkaniuk @marseel - can you take another look (at high-level I'm mostly fine with it)

File: clusterloader2/testing/l4lb/service.yaml
Comment (prameshj) [reply clusterloader2/testing/l4lb/service.yaml:16]: port:8080 and targetPort: 80 should be ok as well, since the ServiceCreationLatency measurement will reach out to VIP:8080 in that case - https://github.com/kubernetes/perf-tests/blob/bb0129087ee1ac79d1800cad913e7f8a252e059b/clusterloader2/pkg/measurement/common/service_creation_latency.go#L340

**Review by prameshj — APPROVED** [review]

File: clusterloader2/pkg/measurement/common/loadbalancer_nodesync_latency.go
Comment (freehan) [reply clusterloader2/pkg/measurement/common/loadbalancer_nodesync_latency.go:133]: Yes. For classic ILB, this measurement should also work. 
But for one test, it can either run only NetLB or ILB. The current test config cannot run a mix of ILB and NetLB

**Comment by freehan** [comment]
Fixed.

File: clusterloader2/pkg/measurement/common/loadbalancer_nodesync_latency.go
Comment (jkaniuk) [inline clusterloader2/pkg/measurement/common/loadbalancer_nodesync_latency.go:142]: Periodic listing of all events could be problematic in the biggest clusters. Could you use approach from [ooms_tracker.go#L125](https://github.com/kubernetes/perf-tests/blob/master/clusterloader2/pkg/measurement/common/ooms_tracker.go#L125) to ignore relists?
It basically ignores initial listing (and listing on e.g. broken watch) and uses only watches with [field selector](https://github.com/kubernetes/perf-tests/blob/master/clusterloader2/pkg/measurement/common/ooms_tracker.go#L203)

Comment (freehan) [reply clusterloader2/pkg/measurement/common/loadbalancer_nodesync_latency.go]: Fixed by using this pattern.

**Comment by jkaniuk** [comment]
/lgtm
/approve
/hold
I would recommend to confirm it actually works in the wild before merging (thus holding), but leaving that up to you.

**Comment by freehan** [comment]
Yes. I have run it many times. THanks!

/hold cancel

**Pull request merged**
